Sample.java

```diff
@@ -1,6 +1,169 @@
+
+    // Time Complexity : O(n)
+    // Space Complexity : O(n)
+    // Did this code successfully run on Leetcode : Yes
+
+    // Your code here along with comments explaining your approach
+    /*
+    Problem 1: 
+    [1,2,3,4] => 1, 1, 2, 6
+                 24,12,4, 1
+    1. Calculate cumulative product from left to right and store it an array.
+    2. Similarly calculate from right to left. (As described in above example)
+    3. Get product of both the arrays
+    */
+
+    public int[] productExceptSelf(int[] nums) {
+
+        if(nums == null || nums.length == 0)
+            return nums;
+            
+        int size = nums.length;
+        int left[] = new int[nums.length];
+
+        left[0] = 1;
+        for(int i = 1; i<size; i++) {
+            left[i] = left[i-1] * nums[i-1];
+        }
+
+        int right[] = new int[nums.length];
+
+        right[size-1] = 1;
+        for(int i = size-2; i>=0; i--) {
+            right[i] = right[i+1] * nums[i+1];
+        }
+
+        int[] res = new int[size];
+
+        for(int i=0; i<size; i++) {
+            res[i] = left[i] * right[i];
+        }
+
+        return res;
+    }
+
+    // Time Complexity : O(N*M)
+    // Space Complexity : O(N*M)
+    // Did this code successfully run on Leetcode : Yes
+
+    // Your code here along with comments explaining your approach
+    /*
+    Problem 2
+    1. Diagonal elements with index position (i, j) have sum of (i+j).
+    2. Map the elements according to the sum.
+    3. Add into final array as per the order. e.g. 1st, 3rd, 5th diagonal elements are in reverse order while 2nd,
+    4th are regular order.
+    */
+
+    public int[] findDiagonalOrder(int[][] matrix) {
+        
+        if(matrix == null || matrix.length == 0){
+            return new int[]{};
+        }
+        
+        int m = matrix.length;
+        int n = matrix[0].length;
+        
+        int max = (m-1)+(n-1);
+        
+        Map<Integer, List<Integer>> map = new HashMap<>();
+        
+        for(int i=0; i<m; i++) {
+            for(int j=0; j<n; j++) {
+                int index = i+j;
+                if(!map.containsKey(index)) {
+                    map.put(index, new ArrayList<Integer>());
+                }
+                
+                map.get(index).add(matrix[i][j]);
+            }
+        }
+        
+        int[] res = new int[m*n];
+        int k = 0;
+        
+        for(int i=0; i<=max; i++) {
+            
+            List<Integer> list = map.get(i);
+            
+            if(i%2 == 0) {
+                for(int j=list.size()-1; j>=0; j--){
+                    res[k] = list.get(j);
+                    k++;
+                }
+            }
+            else {
+                for(int j=0; j<list.size(); j++){
+                    res[k] = list.get(j);
+                    k++;
+                }
+            }
+            
+        }
+        
+        return res;
+        
+    }
+
+    // Time Complexity : O(N)
```

Comment:
Please mention what N is here. Because we have rows and columns in matrix, would be O(RC) as we have to iterate through every element. If N is number of elements in matrix, then it is correct.

Reply:
Yes N is number of elements. I have updated the time complexity